### PR TITLE
Added video timeline view.

### PR DESCRIPTION
**Pull request by dsjen**
screen shot 2015-05-15 at 1 42 22 pm

@brianhw @mulby @jab5569

### CODE REVIEW

File: analytics_dashboard/courses/presenters/engagement.py [hunk 1]
```diff
@@ -203,3 +205,60 @@ def module_id_to_data_id(self, module):
...
+        video_duration = video_module['duration']
+        api_response = self._fill_video_timeline_gaps(api_response, self._calculate_total_video_segments(
+            segment_length, video_duration))
+        return self._build_video_timeline(api_response, segment_length, video_duration)
+
+    def _calculate_total_video_segments(self, segment_length, video_duration=None):
+        """ Return the total number of segments based on video duration or None if duration not specified. """
+        total_segments = None
+        if video_duration:
+            total_segments = int(math.ceil(video_duration/segment_length))
```
Comment (brianhw) [inline analytics_dashboard/courses/presenters/engagement.py]: I'm not sure this is quite right -- or at least, it doesn't quite match the pipeline.  We're just using segment_index = int(duration)/segment_length.  So a duration of 45, say, will be assigned to segment 10.   Your total_segments would say that there were 9 total segments.  Hmm.

File: analytics_dashboard/courses/presenters/engagement.py [hunk 2]
```diff
@@ -203,3 +205,60 @@ def module_id_to_data_id(self, module):
...
+        return total_segments
+
+    def _fill_video_timeline_gaps(self, api_response, segment_total=None):
+        # fill gaps at the beginning
+        first_segment = api_response[0]
+        if first_segment['segment'] > 0:
+            for i in range(first_segment['segment']):
+                api_response.insert(i, self._get_default_video_timeline_segment(i))
+
+        if segment_total and len(api_response) > 1:
```
Comment (brianhw) [inline analytics_dashboard/courses/presenters/engagement.py:239]: nit: symmetric comment for the end?

File: analytics_dashboard/courses/presenters/engagement.py [hunk 3]
```diff
@@ -203,3 +205,60 @@ def module_id_to_data_id(self, module):
...
+            next_segment = api_response[-1]['segment'] + 1
+            for i in range(next_segment, segment_total + 1):
+                api_response.append(self._get_default_video_timeline_segment(i))
+
+        api_response = sorted(api_response, key=lambda segment: segment['segment'])
```
Comment (brianhw) [inline analytics_dashboard/courses/presenters/engagement.py]: nit: I guess if we were going to sort, then it doesn't matter if we insert or append previously.  Did the data require sorting?

**Comment by brianhw** [comment]
Looks good!  Just the one question about mapping duration to segment_id.  Otherwise :+1:

File: analytics_dashboard/courses/presenters/engagement.py (hunk 3 above)
Comment (dsjen) [inline analytics_dashboard/courses/presenters/engagement.py]: Good point!  Sorting isn't needed at this point.  I've removed it.

File: analytics_dashboard/courses/presenters/engagement.py (hunk 2 above)
Comment (dsjen) [inline analytics_dashboard/courses/presenters/engagement.py:239]: It's indeed for the end.  I added a comment.

File: analytics_dashboard/courses/presenters/engagement.py (hunk 1 above)
Comment (dsjen) [inline analytics_dashboard/courses/presenters/engagement.py]: Ah, I see.  I'll update that to `floor` instead.

**Pull request merged**